File: back/api/Mapping/MappedTypes/Product/Product.cs
namespace api.Mapping.MappedTypes.Product;

public record Base
{
  public int Id { get; init; }
  public string Name { get; set; }
  public string Description { get; set; }
  public double Price { get; set; }
  public double DiscountedPrice { get; set; }
  public float DiscountPercent { get; set; }
  public double DiscountAmount { get; set; }
  public bool IsDiscounted { get; set; }
  public bool IsDeleted { get; set; }
  public float AverageStars { get; set; }
  public int ReviewCount { get; set; }
  public int QuestionCount { get; set; }
}

public record ProductResponse : Base
{
  public IEnumerable<ProductBulletPointResponse> BulletPoints { get; init; }
}

public record ShowCaseProductResponse : Base
{
  public IEnumerable<ProductBulletPointResponse> ImportantBulletpoints { get; init; }
}

File: front/types/Product.ts
import { Category } from "./Category";

export type bulletPoint = {
  id: string | null;
  text: string;
};
export const bulletPointDefaultValue = { id: null, text: "" };

export type imageLink = {
  id: string | null;
  link: string;
};
export const imageLinkDefaultValue = { id: null, link: "" };

export type BaseProduct = {
  id: number;
  name: string;
  description: string;
  price: number;
  discountedPrice: number;
  discountPercent: number;
  discountAmount: number;
  isDiscounted: boolean;
  isDeleted: boolean;
  averageStars: number;
  reviewCount: number;
  questionCount: number;
  deepestCategoryId: number;
};

export type Product = BaseProduct & {
  bulletPoints: bulletPoint[];
  images: imageLink[];
};

export type ProductPageProduct = Product & {
  path: Category[];
};
